Add 'child_manipulation_disabled' property to Container

sayaka-sh opened the pull request:
Added the `child_manipulation_disabled` property to the Container class, which stops `queue_sort()` from queueing the sort unless overridden via the `ignore_child_manipulation_disabled` argument.

This allows users to use Containers such as BoxContainers and PanelContainers to place children initially, but then directly control the properties of the children in a script without having to worry about the Container overriding them when the size changes. The issue can somewhat be worked around by setting properties of the children after the Container has sorted them, but this introduces a noticeable flicker.

Example of implementation:
[Without `child_manipulation_disabled`](https://user-images.githubusercontent.com/44363049/129152654-905b689f-f861-422a-88ea-70915723e3f8.mp4)
[With `child_manipulation_disabled`](https://user-images.githubusercontent.com/44363049/129152658-e5f94f55-4e37-44bb-a9ce-22df8d1900f1.mp4)

Master implementation does not work due to https://github.com/godotengine/godot/issues/51543

Edit: Godot improvement proposal [here](https://github.com/godotengine/godot-proposals/issues/3123)
Edit: Renamed property name from `sort_disabled` to `child_manipulation_disabled`

*Bugsquad edit: Closes https://github.com/godotengine/godot-proposals/issues/3123 for `3.x`*

sayaka-sh committed:
Add 'sort_disabled' property to Container

Added the 'sort_disabled' to the Container class, which stops queue_sort() from queueing the sort unless overridden via the 'ignore_sort_disabled' argument.

sayaka-sh committed:
Add 'sort_disabled' property to Container

Added the 'sort_disabled' to the Container class, which stops queue_sort() from queueing the sort unless overridden via the 'ignore_sort_disabled' argument.

groud commented:
I don't think this feature is desired. The whole point of container is to sort their children. So if you don't need sorting, then you should use normal Control nodes instead. Adding an option to disable the sorting of Containers is very confusing IMHO.

spectreseven1138 commented:
> I don't think this feature is desired. The whole point of containers is to sort their children. So if you don't need sorting, then you should use normal Control nodes instead. Adding an option to disable the sorting of Containers is very confusing IMHO.

The sorting is needed to initially place the children or at specific points (such as when adding a new child). The point of this is to be able to control when sorting is enabled.

groud commented:
> The sorting is needed to initially place the children or at specific points (such as when adding a new child). The point of this is to be able to control when sorting is enabled, as it isn't needed all the time.

Well, sorting is the whole point of containers. Disabling sorting is simply a matter of replacing a Container node by a Control node, which is something you could also do at runtime if needed. Consequently, I don't think this feature is needed, and it is very confusing too.

spectreseven1138 commented:
> Well, sorting is the whole point of containers. Disabling sorting is simply a matter of replacing a Container node by a Control node, which is something you could also do at runtime if needed.

You could say the exact same thing about the CollisionShape and CollisionPolygon. You can do that, but it can also introduce other problems like visual flicker and far more complex code within the game code.

> it is very confusing too.

How so? It's essentially the exact same thing as `disable` booleans on other nodes, and if you don't need the property you can just ignore it without having to change anything.

sayaka-sh committed:
Merge branch '3.x-container-sort-disable' of https://github.com/spectreseven1138/godot into 3.x-container-sort-disable

groud commented:
> How so? It's essentially the exact same thing as disable booleans on other nodes, and if you don't need the property you can just ignore it without having to change anything.

Explaining how Control and Containers work is already quite difficult. The two positioning systems (container-based or anchors/offset-based) are right now very separated, and it's a lot easier to explain how they should work by keeping them separated than starting to make situations where containers can work like normal Control nodes. I don't want to have situations where users don't understand why their Container do not work anymore because they checked an option about "sorting" things (sorting being a term quite internal to the Container node).

I'm not 100% against the feature, but in any case, this should have gone through a [Godot Improvement Proposal](https://github.com/godotengine/godot-proposals) first. To see is there are other users who need such a feature, and if there is no confusion about what this does.

spectreseven1138 commented:
> I don't want to have situations where users don't understand why their Container do not work anymore because they checked an option about "sorting" things (sorting being a term quite internal to the Container node).

That makes sense. Maybe a rename to something like `control_mode` or `disable_child_manipulation` would help? I'm sure there are better options though.

> this should have gone through a [Godot Improvement Proposal](https://github.com/godotengine/godot-proposals) first

Really? This is my first time doing a Godot PR so I wasn't sure, but from what I could tell I thought it was fine to open a PR without a proposal if I could implement the feature myself. I agree that this could create confusion so I'll go ahead and make one now though, thanks.

groud commented:
> Maybe a rename to something like control_mode or disable_child_manipulation would help? I'm sure there are better options though.

That could work I guess.

> Really? This is my first time doing a Godot PR so I wasn't sure, but from what I could tell I thought it was fine to open a PR without a proposal if I could implement the feature myself. I agree that this could create confusion so I'll go ahead and make one now though, thanks.

We do accept PR right away for bug fixes and for some non-controversial enhancements (like some UX improvements to the editor, things like that), but we need to be very selective when we change a Node's API. Those changes usually mean we need to keep compatibility with further releases, and thus usually implies more maintenance work than other changes. That's why we need to make sure a feature is both desired and in its best implementation before merging anything (that's what the proposals are supposed to help with).

timothyqiu commented:
@spectreseven1138 I think you can achieve a similar effect by `container.remove_and_skip()`.

spectreseven1138 commented:
> We do accept PR right away for bug fixes and for some non-controversial enhancements (like some UX improvements to the editor, things like that), but we need to be very selective when we change a Node's API. Those changes usually mean we need to keep compatibility with further releases, and thus usually implies more maintenance work than other changes. That's why we need to make sure a feature is both desired and in its best implementation before merging anything (that's what the proposals are supposed to help with).

Thanks for the advice. I opened an improvement proposal [here](https://github.com/godotengine/godot-proposals/issues/3123).

timothyqiu commented:
> especially when I need to move each child back and forth between a Container and non-Container multiple times

I think maybe you misunderstood `remove_and_skip()`. It removes the container itself from the scene tree hierarchy, not its child. You just need to call it in `_ready` and your child controls are not constrainted by the container anymore.

spectreseven1138 commented:
> I think maybe you misunderstood `remove_and_skip()`. It removes the container itself from the scene tree hierarchy, not its child. You just need to call it in `_ready` and your child controls are not constrainted by the container anymore.

It removes the container then moves its children to its parent right? That would mean in order to move the children back into the same container if I need to resort them, I'd have to add the container back into the tree and then move each child back into the container one-by-one while ensuring that the scale, size, and position remain the same. I think that'd work, but like I said it would be far more complex and prone to breaking than being able to just disable the container.

KoBeWi commented:
You could just change the type of the container to Control. Everything stays in place and it's easy to change it back.

timothyqiu commented:
@spectreseven1138 I thought your use case is to only sort the children initially :P

For the case in your videos, you could use a different approach. Instead of changing the position of your buttons, you can wrap your button in an HBoxContainer together with an empty Control node on the left. Changing the min width of that empty Control can achieve what you need, in a more flexible way I think.

[Demo.zip](https://github.com/godotengine/godot/files/6974567/Demo.zip)

spectreseven1138 commented:
> You could just change the type of the container to Control. Everything stays in place and it's easy to change it back.

How do you change node types during runtime? I thought that was impossible.

KoBeWi commented:
Yeah there is no built-in method that does it, but you can easily write one using `Node.replace_by()`.

sayaka-sh committed:
Rename 'sort_disabled' property to 'child_manipulation_disabled'

Rename 'sort_disabled' property to 'child_manipulation_disabled' to make its function clearer

KoBeWi commented:
I recently found a case where this would be useful. If working with inherited/instanced scenes, you aren't able to change node types in the editor. I had a node in a container, but then there was an instance of that scene where it would be useful to disable the container. It's very edge case though and I worked it around by making the instance local.

pycbouh commented:
I don't think this should be accepted. It's a misuse of containers and indeed all the proposed workarounds are available if one really really needs to misuse containers. But this should not be made easier for most users, who should be taught to use the nodes properly instead. We in fact [go to great lengths](https://github.com/godotengine/godot/pull/55157) now to prevent users from changing things about control nodes and positioning in an unsupported way.

akien-mga commented:
Based on the discussion above and in https://github.com/godotengine/godot-proposals/issues/3123, there doesn't seem to be much support for this feature, so closing.

Thanks for contributing nevertheless!
